src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -923,6 +957,835 @@ private FindRequestContext ConstructFindRequestContext(RestRequestContext contex
+                }
+                else
+                {
+                    if (dbResultSetRow is null)
+                    {
+                        throw new DataApiBuilderException(message: "No data returned back from database.",
+                                                          statusCode: HttpStatusCode.InternalServerError,
+                                                          subStatusCode: DataApiBuilderException.SubStatusCodes.DatabaseOperationFailed);
+                    }
+                    else
+                    {
+                        throw new DataApiBuilderException(message: $"Could not insert row with given values for entity: {entityName}",
+                                                          statusCode: HttpStatusCode.Forbidden,
+                                                          subStatusCode: DataApiBuilderException.SubStatusCodes.DatabasePolicyFailure);
+                    }
+                }
+            }
+
+            return dbResultSetRow.Columns;
+        }
+
+        /// <summary>
+        /// Identifies and returns the foreign key definition with the right combination of
+        /// referencing and referenced entity names.
+        /// </summary>
+        /// <param name="sqlMetadataProvider">SqlMetadaProvider object for the given database</param>
+        /// <param name="foreignKeyDefinitions">List of foreign key definitions</param>
+        /// <param name="referencingEntityName">Referencing entity name</param>
+        /// <param name="referencedEntityName">Referenced entity name</param>
+        /// <param name="isMToNRelationship">Indicates whether the relationship is M:N</param>
+        /// <returns>Valid forrign key definition with the right referencing and referencing entities</returns>
+        /// <exception cref="DataApiBuilderException"></exception>
+        private static ForeignKeyDefinition FetchValidForeignKeyDefinition(ISqlMetadataProvider sqlMetadataProvider, List<ForeignKeyDefinition> foreignKeyDefinitions, string referencingEntityName, string referencedEntityName, bool isMToNRelationship)
+        {
+            string referencedTableFullName = sqlMetadataProvider.EntityToDatabaseObject[referencedEntityName].FullName;
+            string referencingTableFullName = sqlMetadataProvider.EntityToDatabaseObject[referencingEntityName].FullName;
+
+            foreach (ForeignKeyDefinition foreignKeyDefinition in foreignKeyDefinitions)
+            {
+                // For a M:N relationship, the referncing table is always the linking table.
+                // So, the referenced table is the only table that needs to be checked.
+                if (isMToNRelationship)
+                {
+                    if (string.Equals(referencedTableFullName, foreignKeyDefinition.Pair.ReferencedDbTable.FullName, StringComparison.OrdinalIgnoreCase))
```

Comment:
The `DatabaseObject` class already has its `Equals` method defined which should ideally be used to compare two objects which is more neater way to compare objects for equality. Let's try use that rather than relying on string comparison.

Reply:
Introduced the method `TryGetFKDefinition` in this PR and added logic to handle M:N relationship